### ISSUE TITLE
how to solve Wrong lane problem perfectly

### ISSUE THREAD

**Issue by thinker310** [eclipse/sumo#11806]
Hello, with the increasing of the number of vehicles in the simulator, the problem that some vehicles enter incorrect lane in the intersection due to routing often happens. Although providing Teleport method in SUMO, the cars cannot be remove or correct immediately, there will be a significant impact on traffic performace evaluation. As shown in the picture below:
![uTools_1665475874220](https://user-images.githubusercontent.com/62609436/195035342-aa64b6ff-b47e-4b29-b6bb-8805e3ebf34d.png)
I wonder if there exist some methods can address the Wrong lane problem. Thank you very much!

**Comment by thinker310** [comment]
> depends on the situation:
> 
> * for a motorway simulation with short on-off-ramps, I recommend this: https://sumo.dlr.de/docs/Simulation/Motorways.html#combined_on-off-ramps
> * for a city scenario it makes sense to configure vehicles with `departLane="best"`
> 
> If neither helps, please attach your scenario files as a zip archive.

[wronglane.zip](https://github.com/eclipse/sumo/files/9764877/wronglane.zip)
I set the departLane but still encountered wrong lane problem. The archive is attached. Thank you!

**Comment by m-kro** [comment]
The "wrong lane" problem is present until v1.14.1 but not anymore in the developer version.

**Comment by thinker310** [comment]
> The "wrong lane" problem is present until v1.14.1 but not anymore in the developer version.

How to get the developer version? I read SUMO documentation but I'm confused. Can I download SUMO files in github to substitute the old files to solve it?

**Comment by m-kro** [comment]
See our website: [SUMO Developement version](https://sumo.dlr.de/docs/Downloads.php#sumo_-_latest_development_version)

**Comment by thinker310** [comment]
> See our website: [SUMO Developement version](https://sumo.dlr.de/docs/Downloads.php#sumo_-_latest_development_version)

Yes, I read it before. It seems that I should git clone repository and install sumo? For simplicity, I would like to download certain files in repository and replace it in my directory to avoid other accident. Can I do it this way?

**Comment by m-kro** [comment]
No, you can't. You have to download either the prebuilt binaries (if available for your system) or build sumo from the source files available on github.

**Comment by thinker310** [comment]
Ok, I see . Thank you!

**Comment by thinker310** [comment]
> No, you can't. You have to download either the prebuilt binaries (if available for your system) or build sumo from the source files available on github.

Excuse me，I installed SUMO developer version but another accident happened as the picture shows.
![uTools_1665631157519](https://user-images.githubusercontent.com/62609436/195491696-58f4998b-8374-458f-8b47-5d6495cf5903.png)
How should I do to solve it?

**Comment by thinker310** [comment]
oh, I think I may have solved it. Thanks!

**Comment by thinker310** [comment]
> The "wrong lane" problem is present until v1.14.1 but not anymore in the developer version.

Unfortunately, I set collision.mingap.factors threshold as your tips and reinstalled SUMO developer version but the wrong lane problem happended again. So is there any solutions now?
![uTools_1665635748661](https://user-images.githubusercontent.com/62609436/195503128-39487989-c1f7-4287-b1c7-7cf09619e7b9.png)
![uTools_1665636053631](https://user-images.githubusercontent.com/62609436/195503210-e30c899e-3ccc-424f-bd06-b90703d39905.png)
![uTools_1665636851886](https://user-images.githubusercontent.com/62609436/195504947-14f53a76-c307-4317-9472-9fea1e1ec32a.png)

**Comment by m-kro** [comment]
@namdre : Could you try to reproduce the result in Linux? I don't get any "wrong lane" issues under Windows.

**Comment by namdre** [comment]
yes but it will probably have to wait until I return from my trip. Also, I don't think whatever happens here should be platform-dependent.

**Comment by thinker310** [comment]
> . I would guess there is some context were are missing

[wrong_lane.zip](https://github.com/eclipse/sumo/files/9788810/wrong_lane.zip)
The case above include all files and I run it on Ubuntu 18.04.

SUMO version as follows:
![uTools_1665769221448](https://user-images.githubusercontent.com/62609436/195908243-71851553-82a2-4b93-85f5-4980592eb6c3.png)
The result remains as before:

![uTools_1665769309841](https://user-images.githubusercontent.com/62609436/195908460-b1bce472-1ddd-464a-a8c0-d841e5b821e9.png)

It's weired, when running it on windows there is no wrong lane problem.

**Comment by thinker310** [comment]
Hello, have you fixed the wrong lane probem? I found a new phenomenon on wrong lane problem like the picture below:
 
![uTools_1666682278028](https://user-images.githubusercontent.com/62609436/197709155-36040052-98a4-4f32-9ba0-eaf1da335b6a.png)
On the north approach, the yellow car wants to go right side and the blue car want to go left side, so they both wrong lane and stuck on the lane, this will make cars behind them congested.
This little problem is really annoying.    (-_-)
Hope it can be refactored as soon as possible.

**Comment by namdre** [comment]
I can confirm the problem: Automatic rerouting does not consider lane changing issues. A car that has already reached the stop line (i.e. halting at a red light) might switch to a route that requires changing to another turn lane. If that lane is already occupied, it will create a blockage. If it happens for two cars that would need to swap their lanes, the blockage is even worse.

The problem is deep-seated because routing doesn't really consider the potential cost of lane changing.  
There is some consideration for not trying to change lanes while close to a junction (`MSVehicle::getRerouteOrigin`) but this concerns brakeGap and so does not take affect for vehicles that are already halting in a queue.

**Comment by thinker310** [comment]
> 

Ok, I have an idea for this problem. Could you change the logic of automatic routing simply? Currently, there is a time interval for impementing vehicle routing, this setting could make the position of routing stochastic, thus the vehicle closed to the intersection may receive routing command and easily lead to wrong lane problem. My solution is to replace time interval with position, e.g. set a position on the lane or edge, every time a car passes the setting position, the routing happend once. In this way, we can set a position far from intersection to ensure the car can change lane without any error.

**Comment by namdre** [comment]
> > 
> 
> Ok, I have an idea for this problem. Could you change the logic of automatic routing simply? Currently, there is a time interval for impementing vehicle routing, this setting could make the position of routing stochastic, thus the vehicle closed to the intersection may receive routing command and easily lead to wrong lane problem. My solution is to replace time interval with position, e.g. set a position on the lane or edge, every time a car passes the setting position, the routing happend once. In this way, we can set a position far from intersection to ensure the car can change lane without any error.

We have something like this already: it's called a rerouter: https://sumo.dlr.de/docs/Simulation/Rerouter.html
